### ISSUE TITLE
Improve flat-field correction

### ISSUE THREAD

**Issue by tischi** [hci-unihd/batchlib#100]
@metavibor @imagirom 

1. I think there is a technical reason why taking the plate mean of the signal may be better than the Chroma slide. The point is that the signal from the sample comes from a relatively thin slice in 3D whereas the Chroma slide is very thick and may thus not give the same profile (one would have to think more about it, but I think that was a point. I could investigate more...) 
@imagirom could you post here a download link to (i) the image from the Chroma slide and (ii) for one plate, the one that you get when you just compute the plate mean of all images? Then I could also have a look.
2. Unfortunately, also for the flat-field formula we need, again, a good background value. @imagirom what value are you using right now?
3. Regarding (2): In this case we have a ground truth that we want to reach: if we apply the formula to all images in the plate and average them again it really should look flat. Either one can use this to analytically solve what we should use or, if the problem cannot be inverted, iteratively optimise the background value until the result is flat.

**Comment by tischi** [comment]
@imagirom @constantinpape @metavibor
Yes, high priority but I think we should **_first_** rerun the analysis without touching the flatfield and just subtracting a fixed background value of 1400 (see Slack) to have some consistent data to work with. My feeling is that, even though it will be good to have an optimal flat-field correction that this should not change the final scores as much...

**Comment by tischi** [comment]
> Regarding (2): In this case we have a ground truth that we want to reach: if we apply the formula to all images in the plate and average them again it really should look flat. Either one can use this to analytically solve what we should use or, if the problem cannot be inverted, iteratively optimise the background value until the result is flat.

It seems like that there is a flaw in this thinking (or I made a mistake in below calculations). But looks like _by definition_ the average corrected signal will always be flat, no matter what we chose as an offset?!

![IMG_1775](https://user-images.githubusercontent.com/2157566/81388529-b4654600-9118-11ea-8809-9d9109b158f9.JPG)

**Comment by constantinpape** [comment]
> Yes, high priority but I think now we could also first rerun the analysis without touching the flatfield and just subtracting a fixed background value of 1400 (see Slack) to have some consistent data to work with. My feeling is that, even though it will be good to have an optimal flat-field correction that this should not change the final scores as much...

Makes sense, we should figure out which plates to rerun for this, but I will move this discussion to slack. (Just in general, I mark everything with `high priority` that should be done for the preprint.)

> (I know that I agreed in the other issue with Vibor that in principle one should first fix the flat-field correction, but I am also a bit worried about @imagirom being overloaded with too many things to do in parallel and not getting enough sleep...)

Yes, I agree we should not change too many things at once, especially if it's things that @imagirom has to do, because he clearly has the highest work-load for the project right now.

**Comment by imagirom** [comment]
> * Unfortunately, also for the flat-field formula we need, again, a good background value. @imagirom what value are you using right now?

You mean to compute the multiplicative part of the flatfield correction? I use the constant camera offset of 550.


> It seems like that there is a flaw in this thinking (or I made a mistake in below calculations). But looks like _by definition_ the average corrected signal will always be flat, no matter what we chose as an offset?!

I think this makes sense, but the contrast will not be constant over the image if we use an incorrect background value

> @imagirom could you post here a download link to (i) the image from the Chroma slide and (ii) for one plate, the one that you get when you just compute the plate mean of all images? Then I could also have a look.

Will do after finishing the PR for ignoring the BG of outliers and making it possible to just set a constant value. If you want to look at the Chroma slides now, they are appended to some email from Vibor.

**Comment by tischi** [comment]
> but the contrast will not be constant over the image if we use an incorrect background value

Yes, I think you are right, so mathematically we would also like to achieve something where `abs( c(x,y) - c(x+dx, y+dy))` is constant across the corrected image `c(x,y)`?! Or how would one express contrast? Anyway, probably getting carried away here.

**Comment by imagirom** [comment]
I think the correct thing might be the following: Additionally to the per-pixel averages over all the images, we could also compute per-pixel variances. Then normalize in a way that makes this constant as well. Maybe `corrected(x, y) = (signal(x, y) - mean(x, y)) / variance(x, y)`.

**Comment by imagirom** [comment]
I will be off for a few hours now, but here is a preliminary result: ratio of pixelwise average over images in plate 311 to current barrel corrector: 
![image](https://user-images.githubusercontent.com/7648199/81426561-a84da880-9159-11ea-9ae3-7056b97d2d21.png)
I think we can definitively improve the barrel correction by going back to the averages over images.

**Comment by tischi** [comment]
> I think we can definitively improve the barrel correction by going back to the averages over images.

Good! As mentioned there also is a microscopy reason why this should be better. Only caveat of this method is that one needs enough images with dense enough cells to cover all pixels nicely, but I think we are in pretty good shape there.

**Comment by imagirom** [comment]
Ok, I updated the barrel correction, now its using the average of all non-outlier images per channel.

**Comment by tischi** [comment]
@imagirom 
Looks much better!
In H01 well (I just looked in one plate) it looks like one could still correct a bit more.
As said, I am confused by my own calculations above, but my intuition is that choosing a value higher than 550 for the offset should increase the correction strength. 

1. If you are motivated you could try for which offset value you get the best results in well H01.
2. Also, I remembered two more relevant publications: 
- https://www.researchgate.net/publication/273637798_CIDRE_An_illumination-correction_method_for_optical_microscopy
- https://www.researchgate.net/publication/317971590_A_BaSiC_tool_for_background_and_shading_correction_of_optical_microscopy_images

If you are interested you could have a look there for more inspirations.

**Comment by imagirom** [comment]
Thanks for the references @tischi. I will have a look.

**Comment by constantinpape** [comment]
We decided to leave this as is for the pre-print, the current version looks good.
